### ISSUE TITLE
[Question] Callback `_on_step` before environment reset

### ISSUE THREAD

**Issue by 19kdc3** [DLR-RM/stable-baselines3#1633]
### ❓ Question

Hi,
I'm trying to use the `_on_step` callback function to save images of my environment after every step. However, I've noticed that on the last step, which then triggers the environment to reset, that the `_on_step` function is ran after the reset, not after the step. This prohibits me from capturing the final step the agent took in my callback and simply returns the freshly reset environment. 

**Current behavior:**
1. Step
2. Reset
3. `_on_step` callback

**Expected behavior:**
1. Step
2. `_on_step` callback
3. Reset

Is there a modification I can do to `_on_step` to have it ran before a reset? or an alternative method to achieve this?
Thanks!

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014)
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by qgallouedec** [comment]
Yes, it's because the environment is vectorized in such a way that the reset is done inside the step. But you can still access the last observation with `infos["terminal_observation"]`.

Example:


```python
from stable_baselines3 import PPO
from stable_baselines3.common.callbacks import BaseCallback

class CustomCallback(BaseCallback):
    def _on_step(self) -> bool:
        for env_idx in range(self.model.n_envs):
            # Get the terminal observation of the just finished trajectory
            last_obs = self.locals["infos"][env_idx].get("terminal_observation")
        return True

PPO("MlpPolicy", "CartPole-v1").learn(10_000, callback=CustomCallback())
```

If you use `render` to "save images" of your environment, I would recommend using a wrapper that saves the rendering in the info dict (eg, with the key "rendering"), so that you can recover it with `self.locals["infos"][env_idx]["rendering"]` (and get the initial rendering with `self.locals["env"].reset_infos["rendering"]`

**Comment by 19kdc3** [comment]
Great! I'll try and implement your suggestion. I'm not familiar with using wrappers for this purpose. Would you be able to provide an example or generally point me in the right direction?

Thanks!

**Comment by araffin** [comment]
> . I'm not familiar with using wrappers for this purpose. Would you be able to provide an example or generally point me in the right direction?

we have tutorial and examples in our doc, and you can look at gymnasium doc too.

**Comment by Yashaswini-Srirangarajan** [comment]
In my custom environment, there are certain variables that take the env.num_timesteps and save multiple files and logs based on the current timestep. This value isn't getting updated on reset since the callback is called only after the reset. How to work around this issue? Thanks!

**Comment by qgallouedec** [comment]
This issue is closed, please open a new issue dedicated to your question, providing an example code and everything
